File: docs/09-Greedy-Algorithms/cpp/activity_selection.cpp
#include <iostream>
#include <vector>
#include <algorithm>

struct Activity {
    int start, end, id;
};

bool compareActivities(Activity a, Activity b) {
    return a.end < b.end;
}

void selectActivities(std::vector<Activity>& activities) {
    std::sort(activities.begin(), activities.end(), compareActivities);
    
    std::cout << "Selected activities:\n";
    // First activity is always selected
    std::cout << "Activity " << activities[0].id << ": (" << activities[0].start << ", " << activities[0].end << ")\n";
    
    int lastEnd = activities[0].end;
    for (size_t i = 1; i < activities.size(); ++i) {
        if (activities[i].start >= lastEnd) {
            std::cout << "Activity " << activities[i].id << ": (" << activities[i].start << ", " << activities[i].end << ")\n";
            lastEnd = activities[i].end;
        }
    }
}

int main() {
    std::vector<Activity> activities = {
        {1, 2, 0}, {3, 4, 1}, {0, 6, 2}, {5, 7, 3}, {8, 9, 4}, {5, 9, 5}
    };
    selectActivities(activities);
    return 0;
}

File: site/09-Greedy-Algorithms/java/ActivitySelection.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

public class ActivitySelection {
    static class Activity {
        int start, end, id;
        Activity(int start, int end, int id) {
            this.start = start;
            this.end = end;
            this.id = id;
        }
    }

    public static void selectActivities(ArrayList<Activity> list) {
        // Sort by end time
        list.sort(Comparator.comparingInt(a -> a.end));

        System.out.println("Selected activities:");
        Activity first = list.get(0);
        System.out.println("Activity " + first.id + ": (" + first.start + ", " + first.end + ")");
        
        int lastEnd = first.end;
        for (int i = 1; i < list.size(); i++) {
            Activity current = list.get(i);
            if (current.start >= lastEnd) {
                System.out.println("Activity " + current.id + ": (" + current.start + ", " + current.end + ")");
                lastEnd = current.end;
            }
        }
    }

    public static void main(String[] args) {
        ArrayList<Activity> activities = new ArrayList<>();
        activities.add(new Activity(1, 2, 0));
        activities.add(new Activity(3, 4, 1));
        activities.add(new Activity(0, 6, 2));
        activities.add(new Activity(5, 7, 3));
        activities.add(new Activity(8, 9, 4));
        activities.add(new Activity(5, 9, 5));

        selectActivities(activities);
    }
}
